frame/contracts/src/wasm/runtime.rs

```diff
@@ -936,243 +937,274 @@ where
 // Any input that leads to a out of bound error (reading or writing) or failing to decode
 // data passed to the supervisor will lead to a trap. This is not documented explicitly
 // for every function.
-define_env!(Env, <E: Ext>,
-	// Account for used gas. Traps if gas used is greater than gas limit.
-	//
-	// NOTE: This is a implementation defined call and is NOT a part of the public API.
-	// This call is supposed to be called only by instrumentation injected code.
-	//
-	// - amount: How much gas is used.
-	[seal0] gas(ctx, amount: u32) => {
+#[define_env(Env)]
+pub mod env {
```

Comment:
We should make this at least magical as possible: If we define something as a module it should also emit a module:

```rust
pub mod <mod_name> {
    struct Env;
    impl x for Env;
}
```

This means the module name is chosen by the user but the emitted struct within it is hard coded. This is no problem as it is scoped under the module. This need of course be documented on the macro.

Reply:
For now it was intentionally done so that the new proc macro expands just the same way the current MBE macro does, to avoid other errors.

If we change it like you suggest, it will require some (probably little) changes in all places where `ImportSatisfyCheck`, `FunctionImplProvider<E>` and `Env` are used, like e.g. 

`struct Env;` ident was parametrized in MBE macro,  which was called with `Test` instead of `Env` in `wasm::prepare` [tests](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/wasm/prepare.rs#L535). 

Still it's definitely not a big deal to make it now expand to some `crate::runtime::<mod_name>` sub-module and change all these places to use it this way (which seems to only be a couple of places in `wasm::prepare` and a single place in [`PrefabWasmMosule::execute()`](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/wasm/mod.rs#L245)

Reply:
It is only used in two places. I think we should make those changes.